portal/migrations/versions/c49b7f7944a7_.py

```diff
@@ -0,0 +1,64 @@
+from alembic import op
+import sqlalchemy as sa
+from sqlalchemy.orm import sessionmaker
+
+from portal.models.audit import Audit
+from portal.models.user_consent import UserConsent
+
+"""consent audit & acceptance_date cleanup
+
+Revision ID: c49b7f7944a7
+Revises: 0f1576a4e220
+Create Date: 2018-06-21 14:37:20.145030
+
+"""
+
+# revision identifiers, used by Alembic.
+revision = 'c49b7f7944a7'
+down_revision = '0f1576a4e220'
+Session = sessionmaker()
+
+
+def upgrade():
+    # ### commands auto generated by Alembic - please adjust! ###
+    op.add_column('user_consents', sa.Column(
+        'acceptance_date', sa.DateTime(), nullable=True))
+
+    bind = op.get_bind()
+    session = Session(bind=bind)
+
+    # Whip through every user_consent, migrating data...
+    delete_em = []
+    for uc in session.query(UserConsent):
+        uc.acceptance_date = uc.audit.timestamp
+
+        # look for `recorded` audit that may have different
+        # timestamp
+        comment = "Consent agreement {} recorded".format(
+            uc.id)
+        recorded = session.query(Audit).filter(
+            Audit.comment == comment)
+        recorded = recorded.first()
+        if recorded:
+            assert(uc.audit.user_id == recorded.user_id)
+            assert(uc.audit.context == 'consent')
+            uc.audit.timestamp = recorded.timestamp
+            delete_em.append(recorded)
+
+    # Purge the now obsolete recorded audit rows
+    [session.delete(r) for r in delete_em]
+    session.commit()
+
+    # Add nullable constraint to `acceptance_date` now that it's populated
+    op.alter_column('user_consents', 'acceptance_date', nullable=False)
+
+
+def downgrade():
+    # Don't see much value in supporting downgrade, so raising for now,
+    # as we'll lose data.
+    #
+    # Those consents w/ differing values in their audit rows from
+    # `acceptance_date` would need recreation of "recorded" audits
+    # in the downgrade.
+    raise RuntimeError(
```

Comment:
This will exception will make it difficult to bisect changes automatically.
I think there's always a risk of losing data when reversing a migration, maybe a logging warning or error instead?

Reply:
if we just log an error, the data is lost, as it'll be allowed to finish.  we want it to fall over.
keep in mind, this is on a `downgrade` only, not part of any rollout process.

Reply:
>keep in mind, this is on a `downgrade` only, not part of any rollout process

That's why I'm not worried about protecting against a `downgrade`. 

The use-case I'm worried about this affecting is using [`git bisect`](https://git-scm.com/docs/git-bisect) to try and determine when a change was introduced. 
If the starting range of commits is large enough, it will span several migrations. In this situation, having a migration to automatically downgrade the DB to match the commit in question makes the bisect process much more "hands off" (and easier to automate).